l10n_br_fiscal/models/document_fiscal_line_mixin_methods.py

```diff
@@ -311,16 +312,80 @@ def _document_comment(self):
+
+        self._onchange_commercial_quantity()
+
+        icms_regulation_id = self.env["l10n_br_fiscal.icms.regulation"]
+
+        if self.product_id.icms_regulation_id:
+            icms_regulation_id = self.product_id.icms_regulation_id
+
+        if not icms_regulation_id:
+            tax_definition_ids = self.env["l10n_br_fiscal.tax.definition"]
+            if self.product_id.ncm_id and self.product_id.ncm_id.tax_definition_ids:
+                tax_definition_ids |= self.product_id.ncm_id.tax_definition_ids
+            if self.product_id.cest_id and self.product_id.cest_id.tax_definition_ids:
+                tax_definition_ids |= self.product_id.cest_id.tax_definition_ids
+            if self.product_id.nbm_id and self.product_id.nbm_id.tax_definition_ids:
+                tax_definition_ids |= self.product_id.nbm_id.tax_definition_ids
+            if self.product_id and self.product_id.tax_definition_ids:
+                tax_definition_ids |= self.product_id.tax_definition_ids
+            if tax_definition_ids:
+                tax_definition_ids = tax_definition_ids.map_tax_definition(
+                    company=self.company_id,
+                    partner=self.partner_id,
+                    product=self.product_id,
+                )
+                icms_regulation_id |= tax_definition_ids.mapped("icms_regulation_id")
+
+        if not icms_regulation_id and self.company_id.icms_regulation_id:
+            icms_regulation_id |= self.company_id.icms_regulation_id
+
+        if not icms_regulation_id and not self.company_id.icms_regulation_id:
+            icms_regulation_id |= self.env.ref("l10n_br_fiscal.tax_icms_regulation")
+
+        if not icms_regulation_id:
+            raise UserError(
+                _(
+                    """Nenhum regulamento do ICMS encontrado,
+                favor revisar a parametrização fiscal"""
+                )
+            )
+
+        if len(icms_regulation_id) > 1:
+            raise UserError(
+                _(
+                    """Multiplos Regulamentos do ICMS encontrados,
+                favor revisar a parametrização fiscal"""
+                )
+            )
+
+        self.icms_regulation_id = icms_regulation_id
+        self._onchange_icms_regulation_id()
+
+    @api.onchange("fiscal_operation_id", "icms_regulation_id", "product_id")
+    def _onchange_icms_regulation_id(self):
+        if not (
+            self.fiscal_operation_id and self.icms_regulation_id and self.product_id
+        ):
+            return
+        self.fiscal_operation_line_id = self.fiscal_operation_id.line_definition(
+            company=self.company_id,
+            partner=self.partner_id,
+            product=self.product_id,
+            icms_regulation=self.icms_regulation_id,
+        )
+
+        if self.fiscal_operation_line_id.force_icms_regulation_id:
+            self.icms_regulation_id = (
+                self.fiscal_operation_line_id.force_icms_regulation_id
```

Comment:
Uma sugestão se não for muito complicado.. eu tenho um pensamento que os onchanges não devem ser muito complexos, será que não é melhor abstrair esse código em outro método ?